### PR TITLE
BUGFIX: Avoid orphaned content nodes when calling publishNode()

### LINKED ISSUE #1617: Orphaned content nodes are left in workspace if calling publishNode()
### Description

When a user modifies a document node and one or more of its content nodes and publishes them using "publish workspace", all these changed nodes are moved to the base workspace and the user's workspace is clean afterwards. If, however, the user publishes only the document node, the content collection nodes and other child nodes of that document remain in the user's workspace. They can be published with "publish workspace", but not individually.

Especially when only content collection nodes remain, this can be quite irritating because no content changes appear anywhere in the user interface, but the user workspace still has changes to be published.

This is somewhat related to https://github.com/neos/neos-development-collection/issues/953.

### Steps to Reproduce

This bug cannot be reproduced using the user interface only because:

1. the content module's publish button will make sure that all changes of the current document are published
2. the workspace module will automatically select all changes of a given document node so that not only the document node can be published individually

The bug can be reproduced by calling the (Neos) Publishing Service's `publishNode()` function by passing only the document node. Therefore this does have an impact on Behat tests, for example, and other applications using the API might leave the content repository in an inconsistent state.

### PR DESCRIPTION
**Pull request by robertlemke**
This changes an issue with using the `PublishingService::publishNode()`
which can result in an inconsistent structure in a user's workspace.

This change also changes the behavior of `PublishingService::discardNode()`
which now will also discard content of a given document node to protect
consistency.

Document nodes and their content must be published or discarded together
in order to protect against inconsistencies when the document node is
moved or removed in one of the base workspaces.

Fixes #1617

### CODE REVIEW

File: TYPO3.Neos/Tests/Behavior/Features/Content/PublishUserWorkspace.feature [hunk 1]
```diff
@@ -7,18 +7,17 @@ Feature: Publish user workspace
...
-      | ecf40ad1-3119-0a43-d02e-55f8b5aa3c70 | /sites               | unstructured              |                   | live      |
-      | fd5ba6e1-4313-b145-1004-dad2f1173a35 | /sites/example       | TYPO3.Neos.NodeTypes:Page | {"title": "Home"} | live      |
+      | Identifier                           | Path           | Node Type                 | Properties        | Workspace |
+      | ecf40ad1-3119-0a43-d02e-55f8b5aa3c70 | /sites         | unstructured              |                   | live      |
+      | fd5ba6e1-4313-b145-1004-dad2f1173a35 | /sites/example | TYPO3.Neos.NodeTypes:Page | {"title": "Home"} | live      |
     And I am authenticated with role "TYPO3.Neos:Editor"
 
   @fixtures
   Scenario: Publish a new ContentCollection with Content
-    Given I am authenticated with role "TYPO3.Neos:Editor"
```
Comment (kdambekalns) [inline TYPO3.Neos/Tests/Behavior/Features/Content/PublishUserWorkspace.feature]: Why is that removed? Or is that a side-effect of decoupling various changes into separate PRs?

Comment (robertlemke) [inline TYPO3.Neos/Tests/Behavior/Features/Content/PublishUserWorkspace.feature]: It's up there: https://github.com/neos/neos-development-collection/pull/1618/files/603c345d0585c277676173425e56ff2ca440a27b#diff-d2ecbf4b70c66889d645ac543d16b216R13

**Comment by robertlemke** [comment]
For the record: I think that the changes in `PublishingService` merely reflect the behavior we established in the user interface (both, Content module and Workspace module). Without the greedy behavior, bugs like https://github.com/neos/neos-development-collection/issues/953 can happen.
